text/0014-network-time.md

```diff
@@ -0,0 +1,683 @@
+
+```csharp
+public struct NetworkTime
+{
+    private int m_TickRate;
+    private float m_TickInterval;
+
+    private int m_Tick;
+    private float m_TickDuration;
+
+    public int Tick => m_Tick;
+
+    public int TickRate => m_TickRate;
+
+    public NetworkTime(int tickRate)
+    {
+        m_TickRate = tickRate;
+        m_TickInterval = 1f / m_TickRate; // potential floating point precision issue, could result in different interval on different machines
+        m_TickDuration = 0;
+        m_Tick = 0;
+    }
+
+    public NetworkTime(int tickRate, int tick, float tickDuration = 0f)
+        : this(tickRate)
+    {
+        Assert.IsTrue(tickDuration < 1f / tickRate);
+
+        m_Tick = tick;
+        m_TickDuration = tickDuration;
+    }
+
+    public NetworkTime(int tickRate, float time)
+        : this(tickRate)
+    {
+        this += time;
+    }
+}
+```
+
+The `NetworkTime` struct will give a user access to similar API functionality as the Unity `Time` API.
+
+```csharp
+public struct NetworkTime
+{
+    //.....
+    public float Time => m_Tick * m_TickInterval + m_TickDuration;
+
+    public float FixedTime => m_Tick * m_TickInterval;
+
+    public float FixedDeltaTime => m_TickInterval;
+
+    public int Tick => m_Tick;
+
+    public int TickRate => m_TickRate;
+    //....
+}
+```
+
+NetworkTime will also provide +- operators for `NetworkTime` and `float` like this:
+
+```csharp
+public static NetworkTime operator +(NetworkTime a, NetworkTime b)
+{
+    Assert.AreEqual(a.TickRate, b.TickRate, $"NetworkTimes must have same TickRate to add.");
+
+    int tick = a.Tick + b.Tick;
+    float tickDuration = a.m_TickDuration + b.m_TickDuration;
+
+    if (tickDuration > a.m_TickInterval)
+    {
+        tick++;
+        tickDuration -= a.m_TickInterval;
+    }
+
+    return new NetworkTime(a.TickRate, tick, tickDuration);
+}
+
+public static NetworkTime operator +(NetworkTime a, float b)
+{
+    a.m_TickDuration += b;
+
+    var deltaTicks = Mathf.FloorToInt(a.m_TickDuration * a.m_TickRate);
+    a.m_TickDuration %= a.m_TickInterval;
+
+    a.m_Tick += deltaTicks;
+
+    return a;
+}
+```
+
+Access to the `NetworkTime` can either be done over the `NetworkManager` for convenience:
+```csharp
+NetworkTime predictedTime = NetworkManager.Singleton.PredictedTime;
+NetworkTime serverTime = NetworkManager.Singleton.ServerTime;
+```
+or by accessing the `PredictedTime` and `ServerTime` properties of the `NetworkTimeSystem` directly.
+
+## Architecture
+
+![architecture](0014-network-time/architecture.png)
+
+- **INetworkTimeProvider** Open interface, in control of how time advances relative to Time.deltaTime.
+- **NetworkTimeSystem** Closed system to keep track of time, convert them to ticks. Also providers helper functions and APIs to interact with network time.
+- **NetworkManager** Executes logic for each network tick. Advances the `NetworkTimeSystem`.
```

Comment:
If I understand this correctly, we're not implementing a new `INetworkUpdateSystem` and instead, `NetworkManager` will tick our `NetworkTickSystem` directly (in other terms, `NetworkUpdateSystem` is not `NetworkUpdateSystem : INetworkUpdateSystem`) right?

Reply:
Correct, the reason why I did it like that is because it made sense to me to have `NetworkTimeSystem` decoupled from the player loop for multiple reasons
- we might want to reuse the time system in a separate scope. Replays, rollback maybe.. haven't thought much about it but I think there could be use cases for creating a separate time system and ticking it.
- gives us more control in the NetworkManager to run code before and after we advance the time system
- easier testing

I'm not very opinionated about this though. If we think that `NetworkTimeSystem` should be a `INetworkUpdateSystem` directly we can do that.

Reply:
good to know. this is more or less an implementation detail — I was just curious, not have an opinion yet.